File: Sources/SomnioServerCore/Connections/WorldClockService.swift
import Foundation
import Logging
import ServiceLifecycle
import SomnioCore
import SomnioData
import SomnioProtocol

/// Drives the in-game world clock at 4× wall clock. Owns the only `WorldClock` instance the
/// server mutates: each tick advances the clock, broadcasts a `DateTick` to every logged-in
/// connection on the configured minute marks, and persists the post-tick state once per
/// in-game minute. The pre-loaded `initialClock` parameter is required (no default seed) so
/// a forgotten pre-load is a compile error rather than a startup race that hands clients
/// `bootDefault` instead of the persisted state.
public actor WorldClockService: Service {
    private let worldRouter: WorldRouter
    private let worldClocks: any WorldClockRepository
    private let interval: Duration
    private let logger: Logger
    private var clock: WorldClock

    public init(
        worldRouter: WorldRouter,
        worldClocks: any WorldClockRepository,
        initialClock: WorldClock,
        interval: Duration = .milliseconds(250),
        logger: Logger
    ) {
        self.worldRouter = worldRouter
        self.worldClocks = worldClocks
        self.clock = initialClock
        self.interval = interval
        self.logger = logger
    }

    public func run() async throws {
        let interval = interval
        let logger = logger
        await cancelWhenGracefulShutdown {
            while !Task.isCancelled {
                do {
                    try await Task.sleep(for: interval)
                } catch is CancellationError {
                    return
                } catch {
                    logger.warning(
                        "world clock sleep failed",
                        metadata: ["error": "\(error)"]
                    )
                    return
                }
                await self.tickOnce()
            }
        }
        // After graceful-shutdown signals: persist whatever the post-tick state is so the
        // last in-game second isn't lost. The service group is already shutting down, so a
        // thrown save error has no recovery path — log and return.
        do {
            try await worldClocks.save(clock)
        } catch {
            logger.warning(
                "world clock final save failed",
                metadata: ["error": "\(error)"]
            )
        }
    }

    /// Direct test seam. Tests drive `tickOnce()` and read `currentTime()` instead of
    /// sleeping the service for `interval`.
    func tickOnce() async {
        let wire = clock.tick()
        // Post-tick second == 0 means we just rolled into a new minute. Two gates run on
        // that boundary: the broadcast gate fires on each mid-hour mark from
        // `SomnioConstants.dateTickMinutes` plus the hour rollover (minute 0); the persist
        // gate fires on every new-minute boundary regardless. The wire's midnight `hour=24`
        // quirk is encoded in `WireTime` so the emit carries it naturally.
        if clock.second == 0 {
            let isMidHourMark = SomnioConstants.dateTickMinutes.contains(clock.minute)
            let isHourRollover = clock.minute == 0
            if isMidHourMark || isHourRollover {
                await worldRouter.broadcastToAllConnections(
                    .dateTick(DateTickMessage(hour: wire.hour, minute: wire.minute))
                )
            }
            do {
                try await worldClocks.save(clock)
            } catch {
                logger.warning(
                    "world clock save failed",
                    metadata: ["error": "\(error)"]
                )
            }
        }
    }

    /// Full clock state for the admin `time` verb readout.
    public func currentTime() -> WorldClock {
        clock
    }

    /// Wire payload used by the per-login and per-portal hooks: a snapshot of the post-tick
    /// internal `(hour, minute)` state. Does not advance the clock and does not reproduce
    /// the midnight `hour=24` quirk (only the advancing tick produces that wire shape, and
    /// these snapshots are post-rollover).
    public func currentDateTickMessage() -> DateTickMessage {
        DateTickMessage(hour: clock.hour, minute: clock.minute)
    }
}

File: packages/server/src/services/worldClockService.ts
import { SOMNIO_CONSTANTS, tickWorldClock } from '@somnio/core'
import type { WorldClock } from '@somnio/core'
import type { DateTickMessage } from '@somnio/protocol'
import type { WorldClockRepository } from '@somnio/data'
import type { Logger } from '../logging.ts'
import type { WorldRouter } from '../world/worldRouter.ts'
import { runPeriodically } from './periodicLoop.ts'

export const DEFAULT_WORLD_CLOCK_INTERVAL_MS = 250

/**
 * The in-game clock at 4x wall clock: each tick advances the clock, broadcasts a `dateTick` on the
 * minute marks and the hour rollover, and persists once per in-game minute. The pre-loaded
 * `initialClock` is required so a forgotten pre-load cannot hand clients the boot default.
 */
export class WorldClockService {
  private readonly worldRouter: WorldRouter
  private readonly worldClocks: WorldClockRepository
  private readonly intervalMs: number
  private readonly logger: Logger
  private readonly clock: WorldClock

  constructor(
    worldRouter: WorldRouter,
    worldClocks: WorldClockRepository,
    initialClock: WorldClock,
    logger: Logger,
    intervalMs: number = DEFAULT_WORLD_CLOCK_INTERVAL_MS
  ) {
    this.worldRouter = worldRouter
    this.worldClocks = worldClocks
    this.clock = { ...initialClock }
    this.logger = logger
    this.intervalMs = intervalMs
  }

  /** Ticks until aborted, then saves whatever the post-tick state is so the last in-game second is not lost. */
  async run(signal: AbortSignal): Promise<void> {
    await runPeriodically(this.intervalMs, signal, () => this.tickOnce())
    try {
      await this.worldClocks.save(this.clock)
    } catch (error) {
      this.logger.warn({ error: String(error) }, 'world clock final save failed')
    }
  }

  /** The test seam: one tick, with the broadcast and persist gates. */
  async tickOnce(): Promise<void> {
    const wire = tickWorldClock(this.clock)
    // A post-tick second of 0 means a new minute: broadcast on the mid-hour marks and the hour
    // rollover, persist regardless. The midnight `hour: 24` quirk rides in `wire`.
    if (this.clock.second !== 0) return
    const minutes: readonly number[] = SOMNIO_CONSTANTS.dateTickMinutes
    if (minutes.includes(this.clock.minute) || this.clock.minute === 0) {
      this.worldRouter.broadcastToAllConnections({
        tag: 'dateTick',
        payload: { hour: wire.hour, minute: wire.minute },
      })
    }
    try {
      await this.worldClocks.save(this.clock)
    } catch (error) {
      this.logger.warn({ error: String(error) }, 'world clock save failed')
    }
  }

  /** Full clock state for the admin `time` verb. */
  currentTime(): WorldClock {
    return { ...this.clock }
  }

  /** The post-tick `(hour, minute)` for the per-login and per-portal hooks; never the midnight 24. */
  currentDateTickMessage(): DateTickMessage {
    return { hour: this.clock.hour, minute: this.clock.minute }
  }
}
